python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+    softmaxes = scorer.final_activation_op(logits_tensor, {})
+    return accuracy(softmaxes, labels_tensor), tf.identity(nll), softmaxes
+
+
+def get_logits_labels(model: tf.keras.Model, evaluation_set: tf.data.TFRecordDataset):
+    """
+    predicts model output on the given evaluation set
+    Returns:
+        model output: numpy.array
+        labels: numpy.array
+    """
+    logits_nps = model.predict(evaluation_set).squeeze()
+    labels_nps = np.concatenate(list(map(lambda x: x[1], evaluation_set))).squeeze().argmax(axis=-1)
+    return logits_nps, labels_nps
+
+
+def temperature_scale(model: tf.keras.Model, scorer: ScorerBase, dataset: RelevanceDataset,
+                      logger: Logger,
+                      logs_dir_local: str, temperature_init: float):
+    """
+    learns a temperature parameter using Temperature Scaling (TS) technique on the validation set.
+    It, then, computes the probability scores of the test set with and without TS and writes them
+    in a .zip file.
+    Parameters:
+        model : tf.keras.Model object to be used for temperature scaling
+        scorer:  ScorerBase object of the RelevanceModel
+        dataset: RelevanceDataset to be used for temperature scaling.
+        logger : Logger
+        logs_dir_local: path to save the TS scores
+        temperature_init: temperature initial value
+    """
+    # get an intermediate model with logits as output --> MUST NOT BE AN ACTIVATION (e.g.
+    # SOFTMAX) LAYER
+    intermediate_model = get_intermediate_model(model, scorer)
+
+    start = tf.Variable(initial_value=[temperature_init], shape=(1), dtype=tf.float32, name="temp",
+                        trainable=True)
+
+    def nll_loss_fn(temperature):
+        """
+        computes Negative Log Likelihood loss by applying temperature scaling.
+        Parameters:
+            temperature: TF.Tensor,  a  parameter of size=(1) to be trained.
+        """
+        logits_tensor = tf.constant(logits_nps, name='logits_tensor', dtype=tf.float32)
+        labels_tensor = tf.constant(labels_nps, name='labels_tensor', dtype=tf.int32)
+        logits_w_temp = tf.divide(logits_tensor, temperature)
+
+        return tf.reduce_mean(
+            tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor,
+                                                           logits=logits_w_temp))
+
+    @tf.function
+    def nll_with_lbfgs():
+        return tfp.optimizer.lbfgs_minimize(make_val_and_grad_fn(nll_loss_fn),
+                                            initial_position=start)
+
+    # evaluation on validation set before temperature scaling
+    logits_nps, labels_nps = get_logits_labels(intermediate_model, dataset.validation)
+    org_acc_op, org_nll_loss_op, _ = eval_relevance_model(scorer, logits_nps,
+                                                          labels_nps)
+
+    # perform temperature scaling
+    results = run(nll_with_lbfgs, logger)
```

Comment:
@marjanhs Any reason we can not use the training set here as well?

Reply:
yes, the calibration must be trained on the validation set (unseen data during training)  then  be evaluated on the test data.

Reply:
Sounds good. Thanks!

Reply:
I thought we have to separate out a _totally_ unseen dataset for calibration?   Meaning: we train the main model on a `train` set, with per-epoch `validation`, then finally express the offline quality on a `test` dataset.  *Then* we need a totally independent `calibration` set, right?

Reply:
@jakemannix  here is the pipeline:
we train  a (classification) model (let's call it model A) using `training set` and validate it on the `validation set`. Actually,
the `validation set` is being used for **training** the temperature parameter during temperature scaling. Here is the reference:[https://github.com/gpleiss/temperature_scaling/blob/master/train.py#L174]( https://github.com/gpleiss/temperature_scaling/blob/master/train.py#L174)
Then, we examine the calibrated model (model B = trained model A + temperature applied on logits) on the `test set` . We save the probability scores of model A and model B of the `test set `to  perform  analysis.